airbyte-integrations/connectors/source-google-search-console/source_google_search_console/streams.py

```diff
@@ -0,0 +1,261 @@
+    """
+
+    data_field = "rows"
+    start_row = 0
+    dimensions = []
+    search_types = ["web", "news", "image", "video"]
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/sites/{stream_slice.get('site_url')}/searchAnalytics/query"
+
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        return "date"
+
+    @property
+    def http_method(self) -> str:
+        return "POST"
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        for site_url in self._site_urls:
+            for search_type in self.search_types:
+                yield {"site_url": site_url, "search_type": search_type}
+
+    def next_page_token(self, response: requests.Response) -> Optional[bool]:
+        if len(response.json().get(self.data_field, [])) == ROW_LIMIT:
+            self.start_row += ROW_LIMIT
+            return True
+
+        self.start_row = 0
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        return {"Content-Type": "application/json"}
+
+    def request_body_json(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Union[Dict[str, Any], str]]:
+
+        start_date = self._start_date
+        if start_date and stream_state:
+            if stream_state.get(self.cursor_field):
+                start_date = max(
+                    pendulum.parse(stream_state[self.cursor_field]),
+                    pendulum.parse(start_date),
+                ).to_date_string()
+
+        data = {
+            "startDate": start_date,
+            "endDate": self._end_date,
+            "dimensions": self.dimensions,
+            "searchType": stream_slice.get("search_type"),
+            "aggregationType": "auto",
+            "startRow": self.start_row,
+            "rowLimit": ROW_LIMIT,
+        }
+        return data
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+        records = response.json().get(self.data_field) or []
+
+        for record in records:
+            record["site_url"] = unquote_plus(stream_slice.get("site_url"))
+            record["searchType"] = stream_slice.get("search_type")
+
+            for dimension in self.dimensions:
+                record[dimension] = record["keys"].pop(0)
+
+            # remove unnecessary empty field
+            record.pop("keys")
+
+            yield record
+
+    def get_updated_state(
```

Comment:
I don't this is correct? it needs to be scoped by searchType. Currently if you sync only 1 search Type then the connector fails, the state output will just say `date: September 1st` which means that we will skip syncing data from previous dates for all other search types. I could be reading this wrong though?

let me know if I can help explain stream slices more clearly, it's definitely not an obvious concept at first

also can you add a unit test for testing state?

Reply:
@sherifnada The functionality of state updates has been updated, but now we are faced with the problem of the acceptance test failing due to nesting, we cannot tell the test to get a value from a nested structure with dynamic fields.
Any suggestions?

Reply:
updated